File: docs/breez-sdk/snippets/python/src/parsing_inputs.py
# pylint: disable=too-many-branches
import logging
from breez_sdk_spark import BreezSdk, InputType, default_config, ExternalInputParser, Network


async def parse_input(sdk: BreezSdk):
    # ANCHOR: parse-inputs
    input_str = "an input to be parsed..."

    try:
        parsed_input = await sdk.parse(input=input_str)
        if isinstance(parsed_input, InputType.BITCOIN_ADDRESS):
            details = parsed_input[0]
            logging.debug(f"Input is Bitcoin address {details.address}")
        elif isinstance(parsed_input, InputType.BOLT11_INVOICE):
            details = parsed_input[0]
            amount = "unknown"
            if details.amount_msat:
                amount = str(details.amount_msat)
            logging.debug(f"Input is BOLT11 invoice for {amount} msats")
        elif isinstance(parsed_input, InputType.LNURL_PAY):
            details = parsed_input[0]
            logging.debug(
                f"Input is LNURL-Pay/Lightning address accepting "
                f"min/max {details.min_sendable}/{details.max_sendable} msats"
            )
        elif isinstance(parsed_input, InputType.LNURL_WITHDRAW):
            details = parsed_input[0]
            logging.debug(
                f"Input is LNURL-Withdraw for min/max "
                f"{details.min_withdrawable}/{details.max_withdrawable} msats"
            )
        elif isinstance(parsed_input, InputType.SPARK_ADDRESS):
            details = parsed_input[0]
            logging.debug(f"Input is Spark address {details.address}")
        elif isinstance(parsed_input, InputType.SPARK_INVOICE):
            invoice = parsed_input[0]
            logging.debug("Input is Spark invoice:")
            if invoice.token_identifier:
                logging.debug(f"  Amount: {invoice.amount} base units of "
                f"token with id {invoice.token_identifier}")
            else:
                logging.debug(f"  Amount: {invoice.amount} sats")

            if invoice.description:
                logging.debug(f"  Description: {invoice.description}")

            if invoice.expires_at:
                logging.debug(f"  Expires at: {invoice.expires_at}")

            if invoice.sender_public_key:
                logging.debug(f"  Sender public key: {invoice.sender_public_key}")
        # Other input types are available
    except Exception as error:
        logging.error(error)
        raise
    # ANCHOR_END: parse-inputs

async def set_external_input_parsers():
    # ANCHOR: set-external-input-parsers
    # Create the default config
    config = default_config(network=Network.MAINNET)
    config.api_key = "<breez api key>"

    # Configure external parsers
    config.external_input_parsers = [
        ExternalInputParser(
            provider_id="provider_a",
            input_regex="^provider_a",
            parser_url="https://parser-domain.com/parser?input=<input>"
        ),
        ExternalInputParser(
            provider_id="provider_b",
            input_regex="^provider_b",
            parser_url="https://parser-domain.com/parser?input=<input>"
        )
    ]
    # ANCHOR_END: set-external-input-parsers

File: docs/breez-sdk/snippets/rust/src/parsing_inputs.rs
use anyhow::Result;
use breez_sdk_spark::*;

async fn parse_input(sdk: &BreezSdk) -> Result<()> {
    // ANCHOR: parse-inputs
    let input = "an input to be parsed...";

    match sdk.parse(input).await? {
        InputType::BitcoinAddress(details) => {
            println!("Input is Bitcoin address {}", details.address);
        }
        InputType::Bolt11Invoice(details) => {
            println!(
                "Input is BOLT11 invoice for {} msats",
                details
                    .amount_msat
                    .map_or("unknown".to_string(), |a| a.to_string())
            );
        }
        InputType::LnurlPay(details) => {
            println!(
                "Input is LNURL-Pay/Lightning address accepting min/max {}/{} msats",
                details.min_sendable, details.max_sendable
            );
        }
        InputType::LnurlWithdraw(details) => {
            println!(
                "Input is LNURL-Withdraw for min/max {}/{} msats",
                details.min_withdrawable, details.max_withdrawable
            );
        }
        InputType::SparkAddress(details) => {
            println!("Input is Spark address {}", details.address);
        }
        InputType::SparkInvoice(invoice) => {
            println!("Input is Spark invoice:");
            if let Some(token_identifier) = &invoice.token_identifier {
                println!(
                    "  Amount: {:?} base units of token with id {}",
                    invoice.amount, token_identifier
                );
            } else {
                println!("  Amount: {:?} sats", invoice.amount);
            }

            if let Some(description) = &invoice.description {
                println!("  Description: {}", description);
            }

            if let Some(expires_at) = invoice.expires_at {
                println!("  Expires at: {}", expires_at);
            }

            if let Some(sender_public_key) = &invoice.sender_public_key {
                println!("  Sender public key: {}", sender_public_key);
            }
        }
        // Other input types are available
        _ => {}
    }
    // ANCHOR_END: parse-inputs
    Ok(())
}

pub(crate) async fn set_external_input_parsers() -> Result<()> {
    // ANCHOR: set-external-input-parsers
    // Create the default config
    let mut config = default_config(Network::Mainnet);
    config.api_key = Some("<breez api key>".to_string());

    // Configure external parsers
    config.external_input_parsers = Some(vec![
        ExternalInputParser {
            provider_id: "provider_a".to_string(),
            input_regex: "^provider_a".to_string(),
            parser_url: "https://parser-domain.com/parser?input=<input>".to_string(),
        },
        ExternalInputParser {
            provider_id: "provider_b".to_string(),
            input_regex: "^provider_b".to_string(),
            parser_url: "https://parser-domain.com/parser?input=<input>".to_string(),
        },
    ]);
    // ANCHOR_END: set-external-input-parsers
    Ok(())
}
